# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

## Split the Data into Training and Testing Sets

In [7]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"]
X = data.drop(columns=["spam"])

print(X.head())
print(y.head())

   word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.00  ...                   0.0 

In [8]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
label_balance = y.value_counts()
print(label_balance)

spam
0    2788
1    1813
Name: count, dtype: int64


In [9]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

X_train shape: (3680, 57)
X_test shape: (921, 57)
y_train shape: (3680,)
y_test shape: (921,)


## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [10]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [11]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)

StandardScaler()

In [12]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
print(X_train_scaled[:5])

[[-0.35849294 -0.1663156  -0.56145245 -0.04622283 -0.46213226 -0.34826528
  -0.28974258 -0.25433765 -0.32332618 -0.37163401 -0.29468554 -0.62650109
  -0.31339295 -0.17532912 -0.19447397 -0.29414289 -0.31752206 -0.35579466
   0.13679517 -0.16618766 -0.67441325 -0.12561726 -0.29347046 -0.215155
  -0.32371654 -0.29991686 -0.22988673 -0.22211804 -0.16692576 -0.21817596
  -0.1695615  -0.14085718 -0.17082404 -0.14304829 -0.18630409 -0.23399564
  -0.32328961 -0.06090721 -0.18036846 -0.18204882 -0.12120056 -0.17268519
  -0.20493605 -0.14140736 -0.30532057 -0.19618137 -0.07194518 -0.1076204
  -0.156207   -0.48491863 -0.15134534 -0.32037559 -0.31155305  0.69767005
  -0.108199   -0.21450276 -0.43523962]
 [-0.08676893 -0.03497293 -0.22384706 -0.04622283  0.74237706 -0.06285915
  -0.28974258  0.38019668 -0.32332618 -0.37163401 -0.29468554  3.24704251
   0.24937939 -0.17532912 -0.19447397 -0.20133663 -0.31752206 -0.35579466
  -0.93262282 -0.16618766 -0.67441325 -0.12561726 -0.29347046 -0.215155
   1

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [13]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(random_state=1)
logistic_model.fit(X_train_scaled, y_train)
X_test_scaled = scaler.transform(X_test)

model_score = logistic_model.score(X_test_scaled, y_test)
print(f"Logistic Regression Model Accuracy: {model_score:.4f}")


Logistic Regression Model Accuracy: 0.9273


In [15]:
# Make and save testing predictions with the saved logistic regression model using the test data
X_test_scaled = scaler.transform(X_test)
y_pred_logistic = logistic_model.predict(X_test_scaled)


# Review the predictions
print(y_pred_logistic[:10])

[0 0 1 0 0 1 1 0 1 1]


In [16]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, y_pred_logistic)
print(f"Accuracy Score: {accuracy:.4f}")

Accuracy Score: 0.9273


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [18]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
random_forest_model = RandomForestClassifier(random_state=1)
random_forest_model.fit(X_train_scaled, y_train)

model_score_rf = random_forest_model.score(X_test_scaled, y_test)
print(f"Random Forest Model Accuracy: {model_score_rf:.4f}")

Random Forest Model Accuracy: 0.9587


In [19]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions_logistic = logistic_model.predict(X_test_scaled)


# Review the predictions
print(testing_predictions_logistic[:10])

[0 0 1 0 0 1 1 0 1 1]


In [20]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_logistic = accuracy_score(y_test, testing_predictions_logistic)
print(f"Logistic Regression Accuracy Score: {accuracy_logistic:.4f}")

Logistic Regression Accuracy Score: 0.9273


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

Which Model Did Better?
After trying both the Logistic Regression and Random Forest models, I saw that the Random Forest model did better. It had an accuracy score of 0.9587, while the Logistic Regression model only had 0.9273.

How Does This Match My Prediction?
I guessed that the Random Forest model would perform better, and it turns out I was right! I think the Random Forest did better because it combines a bunch of decision trees, which helps it find patterns that the Logistic Regression model might miss. The Logistic Regression model is simpler, and maybe it wasn’t enough to handle this dataset as well as the Random Forest could.